File: python/packages/botas/src/botas/state/turn_state.py
"""Turn state container with three scopes: conversation, user, and temp."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional, TypeVar

from botas.state.state_scope import StateScope

if TYPE_CHECKING:
    from botas.core_activity import CoreActivity

T = TypeVar("T")


class TurnState:
    """State container for a single turn with three scopes.

    Provides scoped key-value storage for conversation, user, and temporary state.
    State is loaded at turn start and saved at turn end (if the turn succeeds).

    Example::

        # Access via scopes
        context.state.conversation.set("turnCount", 5)
        count = context.state.conversation.get("turnCount", int)

        # Or via path syntax
        context.state.set_value("conversation.turnCount", 5)
        count = context.state.get_value("conversation.turnCount", int)
    """

    def __init__(
        self,
        activity: "CoreActivity",
        conversation_data: Optional[dict[str, Any]] = None,
        user_data: Optional[dict[str, Any]] = None,
    ) -> None:
        """Initialize turn state with scoped data.

        Args:
            activity: The activity for this turn (used for key derivation).
            conversation_data: Initial conversation scope data.
            user_data: Initial user scope data.
        """
        self._activity = activity
        self._conversation = StateScope(conversation_data)
        self._user = StateScope(user_data)
        self._temp = StateScope()

    @property
    def conversation(self) -> StateScope:
        """Conversation-scoped state (persisted per conversation)."""
        return self._conversation

    @property
    def user(self) -> StateScope:
        """User-scoped state (persisted per user across conversations)."""
        return self._user

    @property
    def temp(self) -> StateScope:
        """Temporary state for the current turn (not persisted)."""
        return self._temp

    def get_value(self, path: str, type_: type[T] = object) -> Optional[T]:
        """Get a value by path.

        Path format: "[scope].property" or "property" (defaults to temp).

        Args:
            path: Dot-separated path (e.g., "conversation.count" or "input").
            type_: Type hint for return value (not enforced at runtime).

        Returns:
            The value if it exists, else None.

        Raises:
            ValueError: If path has more than one dot.
        """
        scope, key = self._parse_path(path)
        return scope.get(key, type_)

    def set_value(self, path: str, value: Any) -> None:
        """Set a value by path.

        Path format: "[scope].property" or "property" (defaults to temp).

        Args:
            path: Dot-separated path (e.g., "conversation.count" or "input").
            value: Value to store.

        Raises:
            ValueError: If path has more than one dot.
        """
        scope, key = self._parse_path(path)
        scope.set(key, value)

    def has_value(self, path: str) -> bool:
        """Check if a value exists at path.

        Args:
            path: Dot-separated path (e.g., "conversation.count" or "input").

        Returns:
            True if the value exists, False otherwise.

        Raises:
            ValueError: If path has more than one dot.
        """
        scope, key = self._parse_path(path)
        return scope.has(key)

    def delete_value(self, path: str) -> None:
        """Delete a value at path.

        Args:
            path: Dot-separated path (e.g., "conversation.count" or "input").

        Raises:
            ValueError: If path has more than one dot.
        """
        scope, key = self._parse_path(path)
        scope.delete(key)

    def delete_conversation_state(self) -> None:
        """Delete all state in the conversation scope."""
        self._conversation.clear()

    def delete_user_state(self) -> None:
        """Delete all state in the user scope."""
        self._user.clear()

    def delete_temp_state(self) -> None:
        """Delete all state in the temp scope."""
        self._temp.clear()

    def _parse_path(self, path: str) -> tuple[StateScope, str]:
        """Parse a path string into (scope, key).

        Args:
            path: Dot-separated path (e.g., "conversation.count" or "input").

        Returns:
            Tuple of (scope object, key string).

        Raises:
            ValueError: If path has more than one dot.
        """
        parts = path.split(".")
        if len(parts) > 2:
            raise ValueError(f"Invalid path: {path} (too many dots)")
        if len(parts) == 1:
            # Unqualified path defaults to temp
            return self._temp, parts[0]
        # Qualified path: "scope.key"
        scope_name, key = parts
        if scope_name == "conversation":
            return self._conversation, key
        if scope_name == "user":
            return self._user, key
        if scope_name == "temp":
            return self._temp, key
        raise ValueError(f"Unknown scope: {scope_name}")

    def get_conversation_key(self) -> str:
        """Derive the storage key for conversation scope from the activity.

        Returns:
            Storage key for conversation state.
        """
        channel_id = self._activity.channel_id or ""
        bot_id = self._activity.recipient.id if self._activity.recipient else ""
        conversation_id = self._activity.conversation.id if self._activity.conversation else ""
        return f"{channel_id}/{bot_id}/conversations/{conversation_id}"

    def get_user_key(self) -> str:
        """Derive the storage key for user scope from the activity.

        Returns:
            Storage key for user state.
        """
        channel_id = self._activity.channel_id or ""
        bot_id = self._activity.recipient.id if self._activity.recipient else ""
        user_id = self._activity.from_account.id if self._activity.from_account else ""
        return f"{channel_id}/{bot_id}/users/{user_id}"

    @staticmethod
    def derive_conversation_key(activity: "CoreActivity") -> str:
        """Derive the storage key for conversation scope from an activity.

        Args:
            activity: The activity to derive the key from.

        Returns:
            Storage key for conversation state.
        """
        channel_id = activity.channel_id or ""
        bot_id = activity.recipient.id if activity.recipient else ""
        conversation_id = activity.conversation.id if activity.conversation else ""
        return f"{channel_id}/{bot_id}/conversations/{conversation_id}"

    @staticmethod
    def derive_user_key(activity: "CoreActivity") -> str:
        """Derive the storage key for user scope from an activity.

        Args:
            activity: The activity to derive the key from.

        Returns:
            Storage key for user state.
        """
        channel_id = activity.channel_id or ""
        bot_id = activity.recipient.id if activity.recipient else ""
        user_id = activity.from_account.id if activity.from_account else ""
        return f"{channel_id}/{bot_id}/users/{user_id}"

File: node/packages/botas-core/src/state/turn-state.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { StateScope } from './state-scope.js'
import { StateScopeImpl } from './state-scope.js'
import type { CoreActivity } from '../core-activity.js'
import type { Storage } from './storage.js'

/**
 * State container for a single turn, providing scoped access to
 * conversation, user, and temporary state.
 *
 * @example
 * ```ts
 * bot.on('message', async (ctx) => {
 *   const count = ctx.state?.conversation.get<number>('count') ?? 0
 *   ctx.state?.conversation.set('count', count + 1)
 *   await ctx.send(`Turn ${count + 1}`)
 * })
 * ```
 */
export interface TurnState {
  /** Conversation-scoped state (persisted per conversation). */
  readonly conversation: StateScope

  /** User-scoped state (persisted per user across conversations). */
  readonly user: StateScope

  /** Temporary state for the current turn (not persisted). */
  readonly temp: StateScope

  /**
   * Get a value by path. Path format: "[scope].property" or "property" (defaults to temp).
   * @example
   * ```ts
   * ctx.state?.getValue<number>('conversation.count')
   * ctx.state?.getValue<string>('input')  // defaults to temp scope
   * ```
   */
  getValue<T = unknown>(path: string): T | undefined

  /**
   * Set a value by path. Path format: "[scope].property" or "property" (defaults to temp).
   * @example
   * ```ts
   * ctx.state?.setValue('conversation.count', 42)
   * ctx.state?.setValue('input', 'hello')  // defaults to temp scope
   * ```
   */
  setValue<T>(path: string, value: T): void

  /**
   * Check if a value exists at path.
   * @example
   * ```ts
   * if (ctx.state?.hasValue('conversation.count')) { ... }
   * ```
   */
  hasValue(path: string): boolean

  /**
   * Delete a value at path.
   * @example
   * ```ts
   * ctx.state?.deleteValue('conversation.count')
   * ```
   */
  deleteValue(path: string): void

  /** Delete all state in the conversation scope. */
  deleteConversationState(): void

  /** Delete all state in the user scope. */
  deleteUserState(): void

  /** Delete all state in the temp scope. */
  deleteTempState(): void
}

/**
 * Internal TurnState implementation.
 * @internal
 */
export class TurnStateImpl implements TurnState {
  readonly conversation: StateScopeImpl
  readonly user: StateScopeImpl
  readonly temp: StateScopeImpl

  private readonly conversationKey: string
  private readonly userKey: string
  private readonly storage: Storage

  private readonly initialConversationHash: string
  private readonly initialUserHash: string
  private conversationDeleted = false
  private userDeleted = false

  constructor(
    storage: Storage,
    conversationKey: string,
    userKey: string,
    conversationData: Record<string, unknown>,
    userData: Record<string, unknown>
  ) {
    this.storage = storage
    this.conversationKey = conversationKey
    this.userKey = userKey

    this.conversation = new StateScopeImpl(conversationData)
    this.user = new StateScopeImpl(userData)
    this.temp = new StateScopeImpl({})

    this.initialConversationHash = this.conversation.getHash()
    this.initialUserHash = this.user.getHash()
  }

  getValue<T = unknown>(path: string): T | undefined {
    const { scope, key } = this.parsePath(path)
    return scope.get<T>(key)
  }

  setValue<T>(path: string, value: T): void {
    const { scope, key } = this.parsePath(path)
    scope.set(key, value)
  }

  hasValue(path: string): boolean {
    const { scope, key } = this.parsePath(path)
    return scope.has(key)
  }

  deleteValue(path: string): void {
    const { scope, key } = this.parsePath(path)
    scope.delete(key)
  }

  deleteConversationState(): void {
    this.conversation.clear()
    this.conversationDeleted = true
  }

  deleteUserState(): void {
    this.user.clear()
    this.userDeleted = true
  }

  deleteTempState(): void {
    this.temp.clear()
  }

  /**
   * Save dirty state to storage.
   * Only writes scopes that have changed since load.
   * @internal
   */
  async saveAsync(): Promise<void> {
    const changes: Record<string, unknown> = {}
    const deletions: string[] = []

    // Check conversation scope
    if (this.conversationDeleted) {
      deletions.push(this.conversationKey)
    } else if (this.conversation.getHash() !== this.initialConversationHash) {
      changes[this.conversationKey] = this.conversation.getData()
    }

    // Check user scope
    if (this.userDeleted) {
      deletions.push(this.userKey)
    } else if (this.user.getHash() !== this.initialUserHash) {
      changes[this.userKey] = this.user.getData()
    }

    // Write and delete in parallel (per spec)
    const promises: Promise<void>[] = []
    if (Object.keys(changes).length > 0) {
      promises.push(this.storage.write(changes))
    }
    if (deletions.length > 0) {
      promises.push(this.storage.delete(deletions))
    }

    await Promise.all(promises)
  }

  /**
   * Parse a path string into scope and key.
   * Path format: "[scope].property" or "property" (defaults to temp).
   * Throws if path has more than one dot.
   */
  private parsePath(path: string): { scope: StateScopeImpl; key: string } {
    const parts = path.split('.')
    
    if (parts.length === 1) {
      // No scope prefix - default to temp
      return { scope: this.temp, key: path }
    }
    
    if (parts.length === 2) {
      const scopeName = parts[0]
      const key = parts[1]
      
      if (!scopeName || !key) {
        throw new Error(`Invalid path: "${path}" - scope and key cannot be empty`)
      }
      
      switch (scopeName.toLowerCase()) {
        case 'conversation':
          return { scope: this.conversation, key }
        case 'user':
          return { scope: this.user, key }
        case 'temp':
          return { scope: this.temp, key }
        default:
          throw new Error(`Invalid scope: "${scopeName}" - must be "conversation", "user", or "temp"`)
      }
    }
    
    throw new Error(`Invalid path: "${path}" - must be "[scope].key" or "key"`)
  }

  /**
   * Create a TurnState from an activity by deriving storage keys.
   * @internal
   */
  static async loadAsync(storage: Storage, activity: CoreActivity): Promise<TurnStateImpl> {
    const conversationKey = deriveConversationKey(activity)
    const userKey = deriveUserKey(activity)

    const results = await storage.read([conversationKey, userKey])
    
    const conversationData = (results[conversationKey] as Record<string, unknown>) ?? {}
    const userData = (results[userKey] as Record<string, unknown>) ?? {}

    return new TurnStateImpl(storage, conversationKey, userKey, conversationData, userData)
  }
}

/**
 * Derive the conversation scope storage key from an activity.
 * Key format: {channelId}/{botId}/conversations/{conversationId}
 * @internal
 */
export function deriveConversationKey(activity: CoreActivity): string {
  const channelId = activity.channelId
  const botId = activity.recipient?.id
  const conversationId = activity.conversation?.id

  if (!channelId || !botId || !conversationId) {
    throw new Error('Activity missing required fields for conversation key: channelId, recipient.id, conversation.id')
  }

  return `${channelId}/${botId}/conversations/${conversationId}`
}

/**
 * Derive the user scope storage key from an activity.
 * Key format: {channelId}/{botId}/users/{userId}
 * @internal
 */
export function deriveUserKey(activity: CoreActivity): string {
  const channelId = activity.channelId
  const botId = activity.recipient?.id
  const userId = activity.from?.id

  if (!channelId || !botId || !userId) {
    throw new Error('Activity missing required fields for user key: channelId, recipient.id, from.id')
  }

  return `${channelId}/${botId}/users/${userId}`
}
